wordpress-mobile/WordPress-iOS
# Auto save published post upon preview and show preview url

Fixes #9287  (also refer to #9282 )
[autosave API call PR on WordPressKit](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/155)

The gist:
When asking to preview a post we check if the post has unsaved changes (now also returning true if the post status is the new status: `AbstractPostRemoteStatusSaved`). 
If the post is not draft, we call the save endpoint that checks if the post blog supports WPRestAPI since this endpoint is not available for self hosted sites.
Not WPCom - we return failure with PostServiceErrorDomain error. and show a **noResultsViewController with "preview not available" - needs copy/ design approval (please see gif)**
Is WPCom - we call the autosave endpoint and get the previewURL on successful response.
We pass the previewURL to the preview generator to display.

WPCom site preview published post:

![preview1](https://user-images.githubusercontent.com/1335657/57185878-3cbb6600-6e89-11e9-8e14-342b8e44c393.gif)

Self hosted site preview published post:
![IMG_4064](https://user-images.githubusercontent.com/1335657/57185884-5066cc80-6e89-11e9-87cd-30dfd173d0a7.GIF)

To test:
 **Following code review and testing by developers, we have discovered an issue where the preview doesn't show the changes for a Jetpack site.**
A. preview post no updates 

1. go to a published post from your post list.
2. go to preview.
3. see the post preview with the latest content.

B. Preview post with updates
1. go to a published post from your post list
2. edit the post
3. preview the post
4. see "Generating preview..."
5. see the post preview with the latest content.

C. preview after discard
1. Steps in B
2. Exit the editor
2. Tap "Discard" when prompted in the action sheet
3. go back to the post
4. hit preview 
5. see the post preview with the latest (published) content.

* I will update the release notes once approved.
